packages/backend/migrations/20240731084359_add_unique_constraint_to_kid.js

```diff
@@ -0,0 +1,29 @@
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.up = function (knex) {
+  // delete any existing duplicates per wallet address
+  knex.raw(`
+  DELETE FROM walletAddressKeys
+    WHERE ctid NOT IN (
+      SELECT MIN(ctid)
+      FROM walletAddressKeys
+      GROUP BY walletAddressId, kid, x
+    );
+  `)
```

Comment:
played around with it and it seems to be working.

What's the user experience going to be for the people using the keys being deleted here? I recall the discussion and read the issue but I guess I'm still not quite connecting the dots. For example, is some normal user going to see `invalid signature` or similar from the grant request process and have no idea they need to upload some new key? Or maybe it just fixes the bug in that issue and otherwise people wont be impacted from these duplicates being deleted? 

I see the error for duplicate writes in gql - that makes sense.

Reply:
The users should not have any issues since they still have another copy of the key. But this makes me think I need to check what happens when a user re-adds a key that they have previously revoked. That should be un-revoked and not throw the duplicate error. And I'll also double check that we actually delete the older versions of the key in the migration and keep the most recent.

Reply:
yeah thats a good point... since we're not deleting these and instead setting `revoked` to true I imagine re-adding the previously revoked key will trigger the unique constraint error. I dont think we can just add the `revoked` to the unique constraint because you couldnt revoke it twice.

1. add key, OK
2. revoke key, OK
3. re-add same key, OK
4. revoke same key, NOT OK. already have a revoked key with these unique fields.

Reply:
Not tested, but I think a [partial index](https://www.postgresql.org/docs/current/indexes-partial.html) like this would work: 

```sql
CREATE UNIQUE INDEX walletAddress_keys_revoked_false_idx
ON "walletAddressKeys" (walletAddressId, kid, x)
WHERE revoked = false;
```

Reply:
do we really want to unrevoke the key? if we do this, we lose the information that sometime in the past, this key was revoked. I think that we need to have new entry in the database if revoked key was again added